### PR TITLE
Feature add custom actions

### PR DESCRIPTION
**Pull request by luc-boussant**

### CODE REVIEW

File: src/adminCore/admin.controller.ts [hunk 1]
```diff
@@ -199,4 +199,20 @@ export class DefaultAdminController {
...
+  async executeListAction(
+    @Req() request: Request,
+    @Body('listActionIndex') listActionIndex: number,
+    @Param() params: AdminModelsQuery,
+    @Response() response: express.Response,
+  ) {
+    const { adminEntity, section, metadata } = await this.getAdminModels(params)
+
+    const listAction = adminEntity.listActions[listActionIndex].action
+    const boundedListAction = listAction.bind(adminEntity)
```
Comment (TheBigBadBunce) [inline src/adminCore/admin.controller.ts:213]: I think you mean `boundListAction` here :)

File: src/adminCore/admin.controller.ts [hunk 2]
```diff
@@ -53,7 +53,7 @@ export class DefaultAdminController {
 
   async getEntityWithRelations(adminEntity: AdminEntity, primaryKey: any) {
     const metadata = adminEntity.metadata
-    const relations = metadata.relations.map(r => r.propertyName)
+    const relations = metadata.relations.map((r) => r.propertyName)
```
Comment (LeoAnesi) [inline src/adminCore/admin.controller.ts:56]: Can we deactivate this rule ? I prefer without the paranthesis

File: src/adminCore/extensions/setAsync.ts [hunk 3]
```diff
@@ -9,7 +9,7 @@
 export class SetAsyncExtension {
   tags = ['setAsync']
 
-  parse = function(parser: any, nodes: any, lexer: any) {
+  parse = function (parser: any, nodes: any, lexer: any) {
```
Comment (LeoAnesi) [inline src/adminCore/extensions/setAsync.ts:12]: This space seems weird to

File: src/public/views/changelist.njk [hunk 4]
```diff
@@ -12,9 +12,19 @@
...
+    {% if adminEntity.listActions %}
+      {% for action in adminEntity.listActions %}
+        <form method="POST" action="{{ 'listAction' | adminUrl(section, metadata) }}">
+          <input type="hidden" name="listActionIndex" value="{{ loop.index0 }}">
+          <button class="btn btn-secondary">{{ action.label }}</button>
+        </form>
+      {% endfor %}
+    {% endif %}
+    <a href="{{ 'add' | adminUrl(section, metadata, entity) }}">
+      <button class="btn btn-primary">Add {{ metadata.name | prettyPrint}}</button>
```
Comment (LeoAnesi) [inline src/public/views/changelist.njk:25]: There is a missing space after `prettyPrint`

File: src/public/scss/changelist.scss [hunk 5]
```diff
@@ -11,6 +11,14 @@
     color: $gray-600;
   }
 
+  &__list-actions {
+    display: flex;
+
+    .btn {
+      margin-left: 0.5rem;
```
Comment (LeoAnesi) [inline src/public/scss/changelist.scss:18]: According to what I read in the rest of the file the convention is bottom + right, can you change it to a right margin elsewhere ?

File: exampleApp/src/user/user.admin.ts [hunk 6]
```diff
@@ -1,12 +1,31 @@
-import { AdminEntity, PasswordWidget } from "nestjs-admin";
-import { User } from "./user.entity";
+import { AdminEntity, DefaultAdminSite, PasswordWidget } from 'nestjs-admin'
+import { Request, Response } from 'express'
+import { User } from './user.entity'
+import { Injectable } from '@nestjs/common'
+import { Repository } from 'typeorm'
+import { InjectRepository } from '@nestjs/typeorm'
 
+@Injectable()
```
Comment (LeoAnesi) [inline exampleApp/src/user/user.admin.ts:8]: Not sure making and entity Injectable is a good practice, would be happy to discuss this with you

File: src/public/scss/changelist.scss (hunk 5 above)
Comment (LeoAnesi) [reply src/public/scss/changelist.scss:18]: Seen together, there is no standard in the code for now, we'll that later

File: src/adminCore/adminEntity.ts [hunk 7]
```diff
@@ -1,11 +1,14 @@
...
+import { EntityType, MaybePromise } from '../types'
 import { getDefaultWidget } from './widgets/utils'
 import DefaultAdminSite from './adminSite'
 import ManyToManyWidget from './widgets/manyToManyWidget'
 import { InvalidDisplayFieldsException } from './exceptions/invalidDisplayFields.exception'
 import { WidgetConstructor } from './widgets/widget.interface'
 
+import { Request, Response } from 'express'
+
+export type ListActionHandler = (request: Request, response: Response) => MaybePromise<void>
```
Comment (LeoAnesi) [inline src/adminCore/adminEntity.ts:11]: Try to use multiple definitions instead of `MaybePromise`

File: src/adminCore/adminEntity.ts [hunk 8]
```diff
@@ -29,13 +29,10 @@ abstract class AdminEntity {
...
 
-  constructor(
-    private readonly adminSite: DefaultAdminSite,
-    private readonly connection: Connection,
-  ) {}
+  constructor(private readonly adminSite: DefaultAdminSite) {}
 
   get repository() {
-    return this.connection.getRepository(this.entity)
+    return this.adminSite.connection.getRepository(this.entity)
```
Comment (LeoAnesi) [inline src/adminCore/adminEntity.ts:48]: The previous commit nead this changes to work, either sqash it split it differently

File: src/adminCore/admin.controller.ts [hunk 9]
```diff
@@ -199,4 +199,20 @@ export class DefaultAdminController {
     )
     return response.redirect(urls.changeListUrl(section, metadata))
   }
+
+  @Post(':sectionName/:entityName/action')
```
Comment (LeoAnesi) [inline src/adminCore/admin.controller.ts:203]: This url is hardcoded as we use generated url from config file in other places

File: exampleApp/src/user/user.admin.ts [hunk 10]
```diff
@@ -28,4 +29,11 @@ export class UserAdmin extends AdminEntity {
     this.userRepository.save(new User())
     request.flash('messages', 'Successfully created')
   }
+
+  async duplicate(entity: User, request: Request, response: Response) {
+    const newUser = new User()
+    newUser.email = entity.email
+    this.userRepository.save(newUser)
+    response.status(200).json({ message: 'It worked' })
+  }
```
Comment (LeoAnesi) [inline exampleApp/src/user/user.admin.ts:38]: split this in a different commit

File: exampleApp/src/user/user.admin.ts (hunk 6 above)
Comment (luc-boussant) [reply exampleApp/src/user/user.admin.ts:8]: As seen together it isn't an entity in our database

File: src/public/views/changelist.njk (hunk 4 above)
Comment (luc-boussant) [reply src/public/views/changelist.njk:25]: done

File: src/adminCore/extensions/setAsync.ts (hunk 3 above)
Comment (luc-boussant) [reply src/adminCore/extensions/setAsync.ts:12]: removed

File: src/adminCore/admin.controller.ts (hunk 2 above)
Comment (luc-boussant) [reply src/adminCore/admin.controller.ts:56]: Yes I set a new prettier config rule

File: src/adminCore/admin.controller.ts (hunk 9 above)
Comment (luc-boussant) [reply src/adminCore/admin.controller.ts:203]: I created a ticket for this one : https://trello.com/c/7ebB8EyK/280-aadev-refactor-the-construction-of-urls-in-the-admincontrollerts-file

File: src/adminCore/adminEntity.ts (hunk 7 above)
Comment (luc-boussant) [reply src/adminCore/adminEntity.ts:11]: I investigated a few possibilities to replace the MaybePromise with multiple definitions without much success 
- I tried to imitate the method overloading properties : https://www.typescriptlang.org/docs/handbook/functions.html#overloads
- I tried using Generics to implement the overloading : https://www.typescriptlang.org/docs/handbook/generics.html
- I tried using an overloaded interface without success also : 
 `export interface ListActionHandler {
  (request: Request, response: Response): Promise<void>
  (request: Request, response: Response): void
}`

I created a ticket if anybody wants to keep investigating : https://trello.com/c/wIDwiJAW/281-aadev-refactor-the-definition-of-the-listactionhandler-and-changelistactionhandler-to-use-overloading-instead-of-maybepromise

File: src/adminCore/adminEntity.ts (hunk 8 above)
Comment (luc-boussant) [reply src/adminCore/adminEntity.ts:48]: Squashed

File: exampleApp/src/user/user.admin.ts (hunk 10 above)
Comment (luc-boussant) [reply exampleApp/src/user/user.admin.ts:38]: Splitted

File: exampleApp/src/app.controller.ts [hunk 11]
```diff
@@ -1,12 +1,12 @@
-import { Controller, Get } from '@nestjs/common';
-import { AppService } from './app.service';
+import { Controller, Get } from '@nestjs/common'
+import { AppService } from './app.service'
```
Comment (LeoAnesi) [inline exampleApp/src/app.controller.ts:2]: Didn't saw that previously, are we sure we want to remove `;` ?

**Review by LeoAnesi — APPROVED** [review]
